### PR TITLE
Forbidden pass context with "beginMemory" or "beginTime" keys to `Profiler::end()` + Validate category + Psalm 1

### PR DESCRIPTION
**Pull request by devanych**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ✔️
| Fixed issues  | -

### CODE REVIEW

**Review by vjik — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,9 +1,9 @@
 # Profiler Change Log
 
-
 ## 1.0.2 under development
 
-- no changes in this release.
+- Chg #42: Forbidden pass context with "beginMemory" or "beginTime" keys to `Profiler::end()` (vjik)
```
Comment (samdark) [inline CHANGELOG.md:5]:
```suggestion
- Bug #42: Forbid passing context with "beginMemory" or "beginTime" keys to `Profiler::end()` (vjik)
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,9 +1,9 @@
...
+- Bug #42: Checking that category of message is a string (vjik)
```
Comment (samdark) [inline CHANGELOG.md:6]:
```suggestion
- Bug #42: Check that category of message is a string (vjik)
```

File: src/Profiler.php [hunk 3]
```diff
@@ -100,28 +105,33 @@ public function getMessages(): array
 
     /**
      * @return TargetInterface[] Profiling targets. Each array element represents
-     * a single {@see TargetInterface profiling target} instance.
+     * a single {@see TargetInterface} profiling target instance.
```
Comment (samdark) [inline src/Profiler.php:108]: It was OK:

image

File: src/Profiler.php [hunk 4]
```diff
@@ -179,6 +190,14 @@ public function end(string $token, array $context = []): void
             }
         }
 
+        if (array_key_exists('beginTime', $context)) {
+            throw new InvalidArgumentException('Forbidden to override "beginTime" in context.');
```
Comment (samdark) [inline src/Profiler.php:194]:
```suggestion
            throw new InvalidArgumentException('It is forbidden to override "beginTime" in context.');
```

File: src/Profiler.php [hunk 5]
```diff
@@ -179,6 +190,14 @@ public function end(string $token, array $context = []): void
...
+        }
+
+        if (array_key_exists('beginMemory', $context)) {
+            throw new InvalidArgumentException('Forbidden to override "beginMemory" in context.');
```
Comment (samdark) [inline src/Profiler.php:198]:
```suggestion
            throw new InvalidArgumentException('It is forbidden to override "beginMemory" in context.');
```

File: tests/ProfilerTest.php [hunk 6]
```diff
@@ -163,11 +163,46 @@ public function testProfileEndWithoutBegin(): void
...
     }
 
+    public function testOverrideBeginTime(): void
+    {
+        $profiler = new Profiler($this->logger);
+        $profiler->begin('test');
+
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectErrorMessage(
+            'Forbidden to override "beginTime" in context.'
```
Comment (samdark) [inline tests/ProfilerTest.php:173]:
```suggestion
            'It is forbidden to override "beginTime" in context.'
```

File: tests/ProfilerTest.php [hunk 7]
```diff
@@ -163,11 +163,46 @@ public function testProfileEndWithoutBegin(): void
...
+    }
+
+    public function testOverrideBeginMemory(): void
+    {
+        $profiler = new Profiler($this->logger);
+        $profiler->begin('test');
+
+        $this->expectException(InvalidArgumentException::class);
+        $this->expectErrorMessage(
+            'Forbidden to override "beginMemory" in context.'
```
Comment (samdark) [inline tests/ProfilerTest.php:185]:
```suggestion
            'It is forbidden to override "beginMemory" in context.'
```

**Review by samdark — CHANGES_REQUESTED** [review]

File: tests/ProfilerTest.php (hunk 7 above)
Comment (vjik) [reply tests/ProfilerTest.php]: done

File: tests/ProfilerTest.php (hunk 6 above)
Comment (vjik) [reply tests/ProfilerTest.php]: done

File: src/Profiler.php (hunk 5 above)
Comment (vjik) [reply src/Profiler.php]: done

File: src/Profiler.php (hunk 4 above)
Comment (vjik) [reply src/Profiler.php]: done

File: CHANGELOG.md (hunk 1 above)
Comment (vjik) [reply CHANGELOG.md]: fixed

File: CHANGELOG.md [hunk 8]
```diff
@@ -1,9 +1,9 @@
 # Profiler Change Log
 
-
 ## 1.0.2 under development
 
-- no changes in this release.
+- Bug #42: Forbid passing context with "beginMemory" or "beginTime" keys to `Profiler::end()` (vjik)
+- Bug #42: Check that category of message is a string (vjik)
```
Comment (samdark) [inline CHANGELOG.md:6]:
```suggestion
- Bug #42: Check that category of a message is a string (vjik)
```

**Review by samdark — APPROVED** [review]

**Comment by samdark** [comment]
:+1

**Pull request merged**